### PR TITLE
Add option to ignore image reference errors

### PR DESCRIPTION
**Pull request by legionus**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1518817

### CODE REVIEW

**Comment by miminar** [comment]
LGTM but I wonder if imageapi.ParseImageStreamTagName and imageapi.ParseImageStreamImageName should be included as well.

**Comment by legionus** [comment]
> I wonder if imageapi.ParseImageStreamTagName and imageapi.ParseImageStreamImageName should be included as well.

@miminar Good catch! So it became even more fun!

File: pkg/oc/admin/prune/images.go [hunk 1]
```diff
@@ -166,6 +167,7 @@ func NewCmdPruneImages(f *clientcmd.Factory, parentName, name string, out io.Wri
...
 	cmd.Flags().BoolVar(&opts.ForceInsecure, "force-insecure", opts.ForceInsecure, "If true, allow an insecure connection to the docker registry that is hosted via HTTP or has an invalid HTTPS certificate. Whenever possible, use --certificate-authority instead of this dangerous option.")
 	cmd.Flags().BoolVar(opts.PruneRegistry, "prune-registry", *opts.PruneRegistry, "If false, the prune operation will clean up image API objects, but the none of the associated content in the registry is removed.  Note, if only image API objects are cleaned up through use of this flag, the only means for subsequently cleaning up registry data corresponding to those image API objects is to employ the 'hard prune' administrative task.")
+	cmd.Flags().BoolVar(&opts.IgnoreInvalidRefs, "ignore-invalid-refs", opts.IgnoreInvalidRefs, "If true, the pruning process will ignore all parser errors of image reference. Usually this means that the prunning process will ignore the connection between the object and the referenced image. As a result the image can be wrongly deleted as unused.")
```
Comment (bparees) [inline pkg/oc/admin/prune/images.go]: grammatical cleanup: "If true, the pruning process will ignore all errors while parsing image references. This means that the pruning process will ignore the intended connection between the object and the referenced image. As a result an image may be incorrectly deleted as unused."

**Comment by legionus** [comment]
@miminar @bparees @dmage please review!

**Comment by miminar** [comment]
LGTM

Deffering to @bparees

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 2]
```diff
@@ -662,6 +668,25 @@ func (p *pruner) addBuildsToGraph(builds *buildapi.BuildList) []error {
 	return errs
 }
 
+// resolveISTagName parses  and tries to find it in the graph. If the parsing fails,
+// an error is returned. If the istag cannot be found, nil is returned.
+func (p *pruner) resolveISTagName(g genericgraph.Graph, referrer *kapi.ObjectReference, istagName string) (*imagegraph.ImageStreamTagNode, error) {
+	name, tag, err := imageapi.ParseImageStreamTagName(istagName)
+	if err != nil {
+		if p.ignoreInvalidRefs {
+			glog.V(4).Infof("Failed to parse ImageStreamTag name %q: %v", istagName, err)
```
Comment (mfojtik) [inline pkg/oc/admin/prune/imageprune/prune.go]: I would make this `glog.Warningf` (guess users should see what they are ignoring).

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 3]
```diff
@@ -687,23 +712,27 @@ func (p *pruner) addBuildStrategyImageReferencesToGraph(referrer *kapi.ObjectRef
 		}
 		ref, err := imageapi.ParseDockerImageReference(from.Name)
 		if err != nil {
-			msg := fmt.Sprintf("failed to parse DockerImage name %q of %s: %v", from.Name, getKindName(referrer), err)
-			glog.V(4).Infof(msg)
-			return []error{newErrBadReferenceToImage(from.Name, referrer, err.Error())}
+			glog.V(4).Infof("failed to parse DockerImage name %q of %s: %v", from.Name, getKindName(referrer), err)
```
Comment (mfojtik) [inline pkg/oc/admin/prune/imageprune/prune.go]: can you capitalize this please? also make it warning as above?

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 2 above)
Comment (legionus) [reply pkg/oc/admin/prune/imageprune/prune.go]: @mfojtik done

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 3 above)
Comment (legionus) [reply pkg/oc/admin/prune/imageprune/prune.go]: @mfojtik done

**Comment by bparees** [comment]
@openshift/cli-review ptal
/approve

File: pkg/oc/admin/prune/imageprune/prune.go [hunk 4]
```diff
@@ -662,6 +668,25 @@ func (p *pruner) addBuildsToGraph(builds *buildapi.BuildList) []error {
...
+			return nil, nil
+		}
+		return nil, newErrBadReferenceTo("ImageStreamTag", istagName, referrer, err.Error())
+	}
+	node := g.Find(imagegraph.ImageStreamTagNodeName(makeISTag(referrer.Namespace, name, tag)))
+	if istNode, ok := node.(*imagegraph.ImageStreamTagNode); ok {
+		return istNode, nil
+	}
+
+	return nil, nil
```
Comment (juanvallejo) [inline pkg/oc/admin/prune/imageprune/prune.go]: would it be worth having an "isNotFound" error or something similar that a caller of this method could check?

**Comment by juanvallejo** [comment]
One comment, otherwise lgtm from a cli perspective

File: pkg/oc/admin/prune/imageprune/prune.go (hunk 4 above)
Comment (legionus) [reply pkg/oc/admin/prune/imageprune/prune.go]: @juanvallejo We return `nil` if nothing is found. It's not an error for this function. Do not need to return an error where it does not exist. The caller function  already handles the `istNode == nil`.

**Pull request merged**